src/EVM/Transaction.hs

```diff
@@ -47,25 +53,63 @@ data Transaction = Transaction {
+                         , ("gas",               (JSON.toJSON $ "0x" ++ showHex (toInteger $ t.txGasLimit) ""))
+                         , ("gasPrice",          (JSON.toJSON $ show $ fromJust $ t.txGasPrice))
+                         , ("v",                 (JSON.toJSON $ show $ (t.txV)-27))
+                         , ("r",                 (JSON.toJSON $ show $ t.txR))
+                         , ("s",                 (JSON.toJSON $ show $ t.txS))
+                         , ("to",                (JSON.toJSON $ t.txToAddr))
+                         , ("nonce",             (JSON.toJSON $ show $ t.txNonce))
+                         , ("value",             (JSON.toJSON $ show $ t.txValue))
+                         , ("type",              (JSON.toJSON $ t.txType))
+                         , ("accessList",        (JSON.toJSON $ t.txAccessList))
+                         , ("maxPriorityFeePerGas", (JSON.toJSON $ show $ fromJust $ t.txMaxPriorityFeeGas))
+                         , ("maxFeePerGas",      (JSON.toJSON $ show $ fromJust $ t.txMaxFeePerGas))
+                         , ("chainId",           (JSON.toJSON ("0x1" :: String))) -- NOTE: should be part of struct?
+                         ]
+
+emptyTransaction :: Transaction
+emptyTransaction = Transaction { txData = mempty
+                               , txGasLimit = 0
+                               , txGasPrice = Nothing
+                               , txNonce = 0
+                               , txR = 0
+                               , txS = 0
+                               , txToAddr = Nothing
+                               , txV = 0
+                               , txValue = 0
+                               , txType = EIP1559Transaction
+                               , txAccessList = []
+                               , txMaxPriorityFeeGas = Nothing
+                               , txMaxFeePerGas = Nothing
+                               }
 
 -- | utility function for getting a more useful representation of accesslistentries
 -- duplicates only matter for gas computation
 txAccessMap :: Transaction -> Map Addr [W256]
 txAccessMap tx = ((Map.fromListWith (++)) . makeTups) tx.txAccessList
   where makeTups = map (\ale -> (ale.accessAddress , ale.accessStorageKeys ))
 
-ecrec :: W256 -> W256 -> W256 -> W256 -> Maybe Addr
-ecrec v r s e = num . word <$> EVM.Precompiled.execute 1 input 32
-  where input = BS.concat (word256Bytes <$> [e, v, r, s])
-
+-- Given chainID and Transaction, it recovers the address that sent it
 sender :: Int -> Transaction -> Maybe Addr
 sender chainId tx = ecrec v' tx.txR  tx.txS hash
   where hash = keccak' (signingData chainId tx)
         v    = tx.txV
         v'   = if v == 27 || v == 28 then v
                else 27 + v
 
+-- We don't know which of 27 or 28 is correct
+-- So we try 28 and try to recover & check if it's correct
+-- if it isn't correct, we need to use the other one
```

Comment:
This comment is outdated now that this logic has been moved into `Sign.hs`

Reply:
Thanks, I removed it. I think it's pretty self-explanatory now, so I think it doesn't need a replacement?